File: elixir/leap/leap.exs
defmodule Year do
  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
    except every year that is evenly divisible by 100
      unless the year is also evenly divisible by 400
  """
  @spec leap_year?(non_neg_integer) :: boolean
  def leap_year?(year) do
    divisible = fn(divisor) -> rem(year, divisor) == 0 end
    divisible.(400) || (divisible.(4) && !divisible.(100))
  end
end

File: ecmascript/leap/leap.js
class Year {
  constructor(year) {
    this.year = year;
  }

  isLeap() {
    const isDivisibleBy = divisor => (this.year % divisor) === 0;
    return isDivisibleBy(400) || (isDivisibleBy(4) && !isDivisibleBy(100));
  }
}

export default Year;
